File: solved/725.SplitLinkedListInParts.cpp
#include <vector> 
#include <iostream> 
using namespace std;

class Solution {
public:
  vector<ListNode*> splitListToParts(ListNode* head, int k) {

    int len = 0;
    for (ListNode* node = head; node != nullptr; node = node->next) {
      len++;
    }

    int elems = len / k; 
    int rem = len % k;

    vector<ListNode*> res;

    for (int listInd = 0; listInd < k; listInd++) {

      int listLen = (listInd < rem) ? (elems+1) : (elems);
      ListNode* newList = nullptr;
      ListNode* last = nullptr;

      for (int size = 0; size < listLen; size++) {
        if (size == 0) {
          newList = head;
          last = head;
        }
        else {
          last->next = head;
          last = head;
        }
        // save next node
        head = head->next;
        // but cut the pointer to the list just generated
        last->next = nullptr;
      }

      res.push_back(newList);
    }

    return res;
  }
};

struct ListNode {
  int val;
  ListNode *next;
  ListNode() : val(0), next(nullptr) {}
  ListNode(int x) : val(x), next(nullptr) {}
  ListNode(int x, ListNode *next) : val(x), next(next) {}
};

File: solved/725.SplitLinkedListInParts.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def splitListToParts(self, head: Optional[ListNode], k: int) -> List[Optional[ListNode]]:

        lenn = 0
        cur = head
        while cur:
            cur = cur.next
            lenn += 1

        size = lenn // k
        rem = lenn % k

        res = []
        start = head
        cur = head
        for _ in range(k):
            cursize = size
            if rem > 0:
                cursize += 1
                rem -= 1

            for _ in range(cursize-1):
                if not cur:
                    break
                cur = cur.next

            res.append(start)
            if cur:
                start = cur.next
                cur.next = None
            else:
                start = None

            cur = start

        return res
